cmd/kubeadm/app/cmd/phases/certs.go

```diff
@@ -188,51 +189,51 @@ func getSANDescription(certSpec *certsphase.KubeadmCert) string {
-		cmd.Flags().StringVar(&cfg.Networking.ServiceSubnet, "service-cidr", cfg.Networking.ServiceSubnet, "Alternative range of IP address for service VIPs, from which derives the internal API server VIP that will be added to the API Server serving cert")
-		cmd.Flags().StringSliceVar(&cfg.APIServerCertSANs, "apiserver-cert-extra-sans", []string{}, "Optional extra altnames to use for the API server serving cert. Can be both IP addresses and DNS names")
-		cmd.Flags().StringVar(&cfg.APIEndpoint.AdvertiseAddress, "apiserver-advertise-address", cfg.APIEndpoint.AdvertiseAddress, "The IP address the API server is accessible on, to use for the API server serving cert")
+func runCAPhase(ca *certsphase.KubeadmCert) func(c workflow.RunData) error {
+	return func(c workflow.RunData) error {
+		data, ok := c.(certsData)
+		if !ok {
+			return errors.New("certs phase invoked with an invalid data struct")
+		}
+
+		if err := certsphase.CreateCACertAndKeyFiles(ca, data.Cfg()); err != nil {
+			return err
+		}
+
+		return nil
 	}
 }
 
-func makeCommandsForCA(ca *certsphase.KubeadmCert, certList []*certsphase.KubeadmCert, cfgPath *string, cfg *kubeadmapiv1beta1.InitConfiguration) []*cobra.Command {
-	subCmds := []*cobra.Command{}
+func makePhasesForCA(ca *certsphase.KubeadmCert, certList []*certsphase.KubeadmCert) []workflow.Phase {
+	subPhases := []workflow.Phase{}
 
-	caCmd := makeCmd(ca, cfgPath, cfg)
-	caCmd.Run = func(cmd *cobra.Command, args []string) {
-		internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(*cfgPath, cfg)
-		kubeadmutil.CheckErr(err)
+	caPhase := makePhase(ca, runCAPhase(ca))
 
-		err = certsphase.CreateCACertAndKeyFiles(ca, internalcfg)
-		kubeadmutil.CheckErr(err)
-	}
-
-	subCmds = append(subCmds, caCmd)
+	subPhases = append(subPhases, caPhase)
 
 	for _, cert := range certList {
-		certCmd := makeCommandForCert(cert, ca, cfgPath, cfg)
-		subCmds = append(subCmds, certCmd)
+		certPhase := makePhase(cert, runCertPhase(cert, ca))
+		subPhases = append(subPhases, certPhase)
 	}
 
-	return subCmds
+	return subPhases
 }
 
-func makeCommandForCert(cert *certsphase.KubeadmCert, caCert *certsphase.KubeadmCert, cfgPath *string, cfg *kubeadmapiv1beta1.InitConfiguration) *cobra.Command {
-	certCmd := makeCmd(cert, cfgPath, cfg)
+func runCertPhase(cert *certsphase.KubeadmCert, caCert *certsphase.KubeadmCert) func(c workflow.RunData) error {
+	return func(c workflow.RunData) error {
+		data, ok := c.(certsData)
+		if !ok {
+			return errors.New("certs phase invoked with an invalid data struct")
+		}
 
-	certCmd.Run = func(cmd *cobra.Command, args []string) {
-		internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(*cfgPath, cfg)
-		kubeadmutil.CheckErr(err)
-		err = configutil.VerifyAPIServerBindAddress(internalcfg.APIEndpoint.AdvertiseAddress)
-		kubeadmutil.CheckErr(err)
+		if err := configutil.VerifyAPIServerBindAddress(data.Cfg().APIEndpoint.AdvertiseAddress); err != nil {
```

Comment:
you can remove the call to `VerifyAPIServerBindAddress` because it is already verified in init.go when creating workflow.RunData

Reply:
Done